### DISCUSSION TITLE
2024/11/09/booleans-are-a-trap/

### DISCUSSION THREAD

**Discussion by giscus** [katafrakt/website_comments#7] [Announcements]
# 2024/11/09/booleans-are-a-trap/

Don't use booleans for domain modeling or at least think twice before doing it.

https://katafrakt.me/2024/11/09/booleans-are-a-trap/

**Comment by sweet2honey** [comment]
Great work! I found it really inspiring. By the way, I noticed a small typo at `Set<PremumFeature>;` :D

**Comment by vevan** [comment]
there are some kinds of doors that can be locked in the open position to avoid accidental closure.

**Comment by 0b00101111** [comment]
I really enjoyed this article! Very insightful approach to domain modeling.

It made me think of an analogy with graph theory. What you're describing is similar to how we represent sparse graphs - using multiple boolean flags is like using an adjacency matrix for a graph where most combinations are invalid (e.g., 12 flags creating 4096 possible states when only 20 are valid). 

Your enum-based approach is more like an adjacency list representation, which is both more efficient and safer since it only explicitly defines valid states and transitions.

**Reply by katafrakt** [reply]
This is very interesting angle to look at it. I very much like it! Thanks for sharing.

**Comment by fortii2** [comment]
Thanks for sharing! That was really insightful!
